internal/controller/magento.go

```diff
@@ -17,23 +17,260 @@ limitations under the License.
+			Named(name).
+			WithOptions(o.ForControllerRuntime()).
+			WithEventFilter(resource.DesiredStateChanged()).
+			For(obj.(client.Object)).
+			Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// A connector is expected to produce an ExternalClient when its Connect method
+// is called.
+type connector struct {
+	kube         client.Client
+	usage        resource.Tracker
+	newServiceFn func(creds []byte, baseURL string) (*MagentoService, error)
+}
+
+// Connect typically produces an ExternalClient by:
+// 1. Tracking that the managed resource is using a ProviderConfig.
+// 2. Getting the managed resource's ProviderConfig.
+// 3. Getting the credentials specified by the ProviderConfig.
+// 4. Using the credentials to form a client.
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	pc := &apisv1alpha1.ProviderConfig{}
+	if err := c.kube.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
+		return nil, errors.Wrap(err, errGetPC)
+	}
+
+	cd := pc.Spec.Credentials
+	data, err := resource.CommonCredentialExtractor(ctx, cd.Source, c.kube, cd.CommonCredentialSelectors)
+	if err != nil {
+		return nil, errors.Wrap(err, errGetCreds)
+	}
+
+	svc, err := c.newServiceFn(data, pc.Spec.MagentoURL)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	client := c.kube
+	return &external{service: svc, kube: client}, nil
+}
+
+// An ExternalClient observes, then either creates, updates, or deletes an
+// external resource to ensure it reflects the managed resource's desired state.
+type external struct {
+	kube client.Client
+	// A 'client' used to connect to the external resource API. In practice this
+	service *MagentoService
+}
+
+// getDesiredCRD returns the CustomResourceDefinition that matches the group and kind.
+func getDesiredCRD(crds *v1.CustomResourceDefinitionList, group string, kind string) *v1.CustomResourceDefinition {
+	for _, crd := range crds.Items {
+		if crd.Spec.Group == group && crd.Spec.Names.Kind == kind {
+			return &crd
+		}
 	}
 	return nil
 }
+
+// Observe checks if the external resource exists and if it is up to date with the managed resource.
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	err := v1.AddToScheme(c.kube.Scheme())
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	crds := &v1.CustomResourceDefinitionList{}
+	_ = c.kube.List(ctx, crds)
+	crd := getDesiredCRD(crds, mg.GetObjectKind().GroupVersionKind().Group, mg.GetObjectKind().GroupVersionKind().Kind)
+	c.service.client.Path = "/rest/" + strings.ToUpper(mg.GetObjectKind().GroupVersionKind().Version) + "/" + crd.Spec.Names.Plural
+
+	externalID := mg.GetAnnotations()["external-id"]
```

Comment:
@ca7alindev package use constant for external-id string, because it used many times.